### PR TITLE
 add template wrappers for ScaLAPACK functions

### PR DESCRIPTION
**Pull request by BenBrands**
ScaLAPACKMatrix should now work for **float** and **double**

added if-clause in **source/base/process_grid.cc** as tests with more than one MPI process used to fail on my local machine

### CODE REVIEW

**Comment by BenBrands** [comment]
ping @davydden

**Comment by masterleinad** [comment]
You didn't mean to close this PR, did you?

**Comment by BenBrands** [comment]
I wanted to squash the commits. If that is not necessary I would be happy to leave it open!

**Comment by masterleinad** [comment]
You can still squash and (force) push to this PR. There is no need to create a new one.

**Comment by BenBrands** [comment]
OK, as I'm not too familiar I did not know that!

**Comment by davydden** [comment]
Ps github does not like brackets, what I meant is ‘test\<double\>()’ and same for float

**Comment by davydden** [comment]
For squashing do:
git rebase -i HEAD~5

(5 commits)

And before “push -f” you can diff with the remote branch to make sure you didn’t accidentally lost some commits.

**Comment by BenBrands** [comment]
Yes, the tests were written in that way. I will add the tests for float on the weekend!

**Comment by BenBrands** [comment]
Adapted the tests to run with float and double!
For floats I had to relax the error tolerances significantly. Test no. 4 is currently running only for double, as the error in the Frobenius norm is unexpectedly larger for floats (@davydden, we can discuss this on Monday).

**Comment by BenBrands** [comment]
Due to squashing and rebasing I have accidentaly mixed the current PR with other things :(
I will take care of that on Monday.

File: tests/scalapack/scalapack_06.cc [hunk 1]
```diff
@@ -40,7 +40,25 @@ extern "C"  //Some Lapack routines
...
 template <typename NumberType>
-void test(const unsigned int size, const unsigned int block_size)
+void syev(char *jobz, char *uplo, int *n, NumberType *A, int *lda, NumberType *w, NumberType *work, int *lwork, int *info)
+{
+	Assert (false, dealii::ExcNotImplemented());
+}
+
+void syev(char *jobz, char *uplo, int *n, float *A, int *lda, float *w, float *work, int *lwork, int *info)
+{
+	ssyev_(jobz,uplo,n,A,lda,w,work,lwork,info);
```
Comment (davydden) [inline tests/scalapack/scalapack_06.cc]: maybe move it to lapack templates? You don't have to use it in wrappers though.

**Comment by davydden** [comment]
@BenBrands please re-run the indentation script.

**Comment by BenBrands** [comment]
@davydden indentation script was executed!
Added copy routine and associated tests and also replaced int by types::blas_int.

File: include/deal.II/lac/scalapack.h [hunk 2]
```diff
@@ -158,6 +158,15 @@ class ScaLAPACKMatrix : protected TransposeTable<NumberType>
    */
   void copy_to (FullMatrix<NumberType> &matrix) const;
 
+
+  /*
+   * Copy the contents of the distributed matrix into different distributed matrix @p dest.
+   * The function is supposed to also work for matrices with different process grids or
```
Comment (davydden) [inline include/deal.II/lac/scalapack.h]: > The function also works for matrices...

File: source/base/process_grid.cc [hunk 3]
```diff
@@ -48,32 +49,32 @@ namespace
 
     // Below we always try to create 2D processor grids:
 
-    int n_processes;
+    dealii::types::blas_int n_processes;
```
Comment (davydden) [inline source/base/process_grid.cc]: that actually should not be changed, sorry I mislead you a bit by saying that one can change **all* `int` to `dealii::types::blas_int`, have not thought about those things:

```
int MPI_Comm_size( MPI_Comm comm, int *size ) 
```

File: source/base/process_grid.cc [hunk 4]
```diff
@@ -148,20 +149,20 @@ namespace Utilities
...
               ExcInternalError());
 
-      std::vector<int> inactive_with_root_ranks;
+      std::vector<dealii::types::blas_int> inactive_with_root_ranks;
       inactive_with_root_ranks.push_back(0);
       for (unsigned int i = n_active_mpi_processes; i < n_mpi_processes; ++i)
         inactive_with_root_ranks.push_back(i);
 
       // Get the group of processes in mpi_communicator
-      int ierr = 0;
```
Comment (davydden) [inline source/base/process_grid.cc]: same here.

File: source/base/process_grid.cc [hunk 5]
```diff
@@ -148,20 +149,20 @@ namespace Utilities
...
+      dealii::types::blas_int ierr = 0;
       MPI_Group all_group;
       ierr = MPI_Comm_group(mpi_communicator, &all_group);
       AssertThrowMPI(ierr);
 
       // Construct the group containing all ranks we need:
       MPI_Group inactive_with_root_group;
-      const int n = inactive_with_root_ranks.size();
```
Comment (davydden) [inline source/base/process_grid.cc]: and here https://www.mpich.org/static/docs/v3.1/www3/MPI_Group_incl.html

File: source/base/process_grid.cc [hunk 6]
```diff
@@ -223,12 +224,12 @@ namespace Utilities
 
 
     template <typename NumberType>
-    void ProcessGrid::send_to_inactive(NumberType *value, const int count) const
+    void ProcessGrid::send_to_inactive(NumberType *value, const dealii::types::blas_int count) const
     {
       Assert (count>0, ExcInternalError());
       if (mpi_communicator_inactive_with_root != MPI_COMM_NULL)
         {
-          const int ierr =
```
Comment (davydden) [inline source/base/process_grid.cc]: and here

File: source/base/process_grid.cc [hunk 7]
```diff
@@ -223,12 +224,12 @@ namespace Utilities
 
 
     template <typename NumberType>
-    void ProcessGrid::send_to_inactive(NumberType *value, const int count) const
+    void ProcessGrid::send_to_inactive(NumberType *value, const dealii::types::blas_int count) const
```
Comment (davydden) [inline source/base/process_grid.cc]: and here: 
```
int MPI_Bcast( void *buffer, int count, MPI_Datatype datatype, int root, 
               MPI_Comm comm )
```

File: source/lac/scalapack.cc [hunk 8]
```diff
@@ -205,6 +205,88 @@ ScaLAPACKMatrix<NumberType>::copy_to (FullMatrix<NumberType> &matrix) const
 
 
 
+/*
```
Comment (davydden) [inline source/lac/scalapack.cc]: you already documented it, no need to put `doxygen` here.

File: source/lac/scalapack.cc [hunk 9]
```diff
@@ -205,6 +205,88 @@ ScaLAPACKMatrix<NumberType>::copy_to (FullMatrix<NumberType> &matrix) const
...
+   * just in case of different process grids or block-cyclic distributions
+   * inter-process communication is necessary
+   * if distributed matrices have the same process grid and block sizes, local copying is enough
+   */
+  if ( (this->grid != dest.grid) || (row_block_size != dest.row_block_size) || (column_block_size != dest.column_block_size) )
+    {
+      /*
+       * get the MPI communicator, which is the union of the source and destination MPI communicator
+       */
+      dealii::types::blas_int ierr = 0;
```
Comment (davydden) [inline source/lac/scalapack.cc]: int

File: source/lac/scalapack.cc [hunk 10]
```diff
@@ -205,6 +205,88 @@ ScaLAPACKMatrix<NumberType>::copy_to (FullMatrix<NumberType> &matrix) const
...
+      MPI_Group group_source, group_dest, group_union;
+      ierr = MPI_Comm_group(this->grid->mpi_communicator, &group_source);
+      AssertThrowMPI(ierr);
+      ierr = MPI_Comm_group(dest.grid->mpi_communicator, &group_dest);
+      AssertThrowMPI(ierr);
+      ierr = MPI_Group_union(group_source, group_dest, &group_union);
+      AssertThrowMPI(ierr);
+      MPI_Comm mpi_communicator_union;
+      ierr = MPI_Comm_create_group(MPI_COMM_WORLD, group_union, 5, &mpi_communicator_union);
```
Comment (davydden) [inline source/lac/scalapack.cc]: we should not hard-code `MPI_COMM_WORLD`. I think the right thing to ask is that both matrices are built upon the same mpi_communicator (`this->grid->mpi_communicator` and `dest.grid->mpi_communicator`) or not?

File: tests/scalapack/scalapack_08.cc [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE at
+// the top level of the deal.II distribution.
+//
+// ---------------------------------------------------------------------
+
+#include "../tests.h"
+#include "../lapack/create_matrix.h"
+
+// test Cholesky decomposition in ScaLAPACK vs FullMatrix
```
Comment (davydden) [inline tests/scalapack/scalapack_08.cc]: please adjust the description.

File: include/deal.II/lac/scalapack.h [hunk 12]
```diff
@@ -158,6 +158,15 @@ class ScaLAPACKMatrix : protected TransposeTable<NumberType>
    */
   void copy_to (FullMatrix<NumberType> &matrix) const;
 
+
+  /*
```
Comment (bangerth) [inline include/deal.II/lac/scalapack.h]: You'll need a second asterisk here for doxygen to pick this up.

File: include/deal.II/lac/scalapack.h [hunk 13]
```diff
@@ -158,6 +158,15 @@ class ScaLAPACKMatrix : protected TransposeTable<NumberType>
...
+   * Copy the contents of the distributed matrix into different distributed matrix @p dest.
```
Comment (bangerth) [inline include/deal.II/lac/scalapack.h]: -> into a differently distributed matrix

File: include/deal.II/lac/scalapack.templates.h [hunk 14]
```diff
@@ -68,62 +68,73 @@ extern "C"
    *
    * https://www.ibm.com/support/knowledgecenter/en/SSNR5K_4.2.0/com.ibm.cluster.pessl.v4r2.pssl100.doc/am6gr_dbint.htm
    */
-  void Cblacs_gridinit(int *context, const char *order, int grid_height, int grid_width);
+  void Cblacs_gridinit(dealii::types::blas_int *context, const char *order, dealii::types::blas_int grid_height,
+                       dealii::types::blas_int grid_width);
```
Comment (bangerth) [inline include/deal.II/lac/scalapack.templates.h]: while you're here already, please split the arguments onto separate lines each

Comment (bangerth) [reply include/deal.II/lac/scalapack.templates.h]: same for the following functions, if you wouldn't mind

File: source/base/process_grid.cc [hunk 15]
```diff
@@ -216,18 +217,19 @@ namespace Utilities
       if (mpi_process_is_active)
         Cblacs_gridexit(blacs_context);
 
-      MPI_Comm_free(&mpi_communicator_inactive_with_root);
+      if (mpi_communicator_inactive_with_root != MPI_COMM_NULL)
```
Comment (bangerth) [inline source/base/process_grid.cc]: I think you need to use `MPI_Comm_compare` for this unless `MPI_COMM_NULL` was the initializer of the variable and you never assigned anything to it.

File: source/base/process_grid.cc [hunk 16]
```diff
@@ -216,18 +217,19 @@ namespace Utilities
...
     template <typename NumberType>
-    void ProcessGrid::send_to_inactive(NumberType *value, const int count) const
+    void ProcessGrid::send_to_inactive(NumberType *value, const dealii::types::blas_int count) const
     {
       Assert (count>0, ExcInternalError());
       if (mpi_communicator_inactive_with_root != MPI_COMM_NULL)
         {
-          const int ierr =
+          const dealii::types::blas_int ierr =
             MPI_Bcast(value,count,
```
Comment (bangerth) [inline source/base/process_grid.cc]: this seems wrong -- MPI_Bcast returns an `int` as return value.

File: source/base/process_grid.cc (hunk 15 above)
Comment (davydden) [reply source/base/process_grid.cc]: @bangerth actually by construction this communicator is either something useful or MPI_COMM_NULL, see https://github.com/dealii/dealii/blob/master/source/base/process_grid.cc#L170-L174
so i think the suggested change is correct.

**Comment by BenBrands** [comment]
Hopefully, all remarks have been adressed!

File: source/base/process_grid.cc (hunk 15 above)
Comment (bangerth) [reply source/base/process_grid.cc]: OK, thanks for confirming!

File: source/base/process_grid.cc (hunk 7 above)
Comment (davydden) [reply source/base/process_grid.cc]: @BenBrands i think you forgot to switch this one back to `int`.

Comment (BenBrands) [reply source/base/process_grid.cc]: Yup, sorry!

**Comment by davydden** [comment]
./run-tests

File: include/deal.II/base/process_grid.h [hunk 17]
```diff
@@ -118,7 +120,7 @@ namespace Utilities
        * are not in the process grid.
        */
       template <typename NumberType>
-      void send_to_inactive(NumberType *value, const int count=1) const;
```
Comment (davydden) [inline include/deal.II/base/process_grid.h]: @BenBrands this still needs to be reverted.

File: source/base/process_grid.cc (hunk 7 above)
Comment (davydden) [reply source/base/process_grid.cc]: and here...

File: source/base/process_grid.cc [hunk 18]
```diff
@@ -242,8 +243,8 @@ namespace Utilities
 
 // instantiations
 
-template void Utilities::MPI::ProcessGrid::send_to_inactive<double>(double *, const int) const;
```
Comment (davydden) [inline source/base/process_grid.cc]: and here

File: source/base/process_grid.cc (hunk 7 above)
Comment (davydden) [reply source/base/process_grid.cc]: ah, it looks like this is the one place. the other two are ok already.

**Comment by davydden** [comment]
@BenBrands here's the patch to fix compiler errors with 64bit blas:
```
diff --git a/source/base/process_grid.cc b/source/base/process_grid.cc
index e210a7d69b..e3480863de 100644
--- a/source/base/process_grid.cc
+++ b/source/base/process_grid.cc
@@ -56,7 +56,7 @@ namespace
     // with rectangular blocks when every core owns only a single block:
     const dealii::types::blas_int n_processes_heuristic = dealii::types::blas_int(std::ceil((1.*m)/block_size_m))*
                                                           dealii::types::blas_int(std::ceil((1.*n)/block_size_n));
-    const dealii::types::blas_int Np = std::min(n_processes_heuristic, n_processes);
+    const dealii::types::blas_int Np = std::min(n_processes_heuristic, dealii::types::blas_int(n_processes));
 
     // Now we need to split Np into  Pr x Pc. Assume we know the shape/ratio
     // Pc =: ratio * Pr
@@ -72,7 +72,7 @@ namespace
     // but this affects the grid shape dramatically, i.e. 10 cores 3x3 becomes 2x5.
 
     // limit our estimate to be in [2, Np]
-    dealii::types::blas_int n_process_columns = std::min (Np, std::max(2, Pc));
+    dealii::types::blas_int n_process_columns = std::min (Np, std::max(dealii::types::blas_int(2), Pc));
     // finally, get the rows:
     dealii::types::blas_int n_process_rows = Np / n_process_columns ;
 
@@ -224,7 +224,7 @@ namespace Utilities
 
 
     template <typename NumberType>
-    void ProcessGrid::send_to_inactive(NumberType *value, const dealii::types::blas_int count) const
+    void ProcessGrid::send_to_inactive(NumberType *value, const int count) const
     {
       Assert (count>0, ExcInternalError());
       if (mpi_communicator_inactive_with_root != MPI_COMM_NULL)
diff --git a/source/lac/scalapack.cc b/source/lac/scalapack.cc
index fc5526ed2a..590b6645d4 100644
--- a/source/lac/scalapack.cc
+++ b/source/lac/scalapack.cc
@@ -62,7 +62,7 @@ ScaLAPACKMatrix<NumberType>::ScaLAPACKMatrix(const size_type n_rows_,
       n_local_columns = numroc_(&n_columns, &column_block_size, &(grid->this_process_column), &first_process_column, &(grid->n_process_columns));
 
       // LLD_A = MAX(1,NUMROC(M_A, MB_A, MYROW, RSRC_A, NPROW)), different between processes
-      dealii::types::blas_int lda = std::max(1,n_local_rows);
+      dealii::types::blas_int lda = std::max(dealii::types::blas_int(1),n_local_rows);
 
       dealii::types::blas_int info=0;
       descinit_(descriptor, &n_rows, &n_columns,
```
please apply it to your branch... I will now check if tests pass...

**Comment by BenBrands** [comment]
OK, I will apply the changes tomorrow!

**Comment by davydden** [comment]
@BenBrands looks like all the process-grid (cblas) related functions are with `int` regardless of what BLAS uses.

**Comment by davydden** [comment]
@BenBrands 
I am afraid we need to leave `dealii::types::blas_int` out of this PR. I simply can't find any info about the ScaLAPACK functions signatures with 64bit BLAS. Everything I can find use `int` throughout. Sorry for setting you on the wrong track with this change.

**Comment by BenBrands** [comment]
@davydden 
No problem, I will revert dealii::types::blas_int to int.

**Comment by BenBrands** [comment]
Done!

**Comment by BenBrands** [comment]
Are there any further suggestions or requests?

**Comment by davydden** [comment]
> Are there any further suggestions or requests?

not from my side, i will merge it tomorrow if there won't be further comments from othera

**Comment by masterleinad** [comment]
In the current version, the tests
- scalapack/scalapack_05
- scalapack/scalapack_06
- scalapack/scalapack_07

 don't finish for me when running with more than one process, i.e. there seem to be some deadlock.

**Comment by masterleinad** [comment]
In `reciprocal_condition_number` the call to `ppocon` seems to require communication with an inactive MPI process.

**Comment by BenBrands** [comment]
@masterleinad: Is that the case for every MPI run with more than one process?

**Comment by davydden** [comment]
> Is that the case for every MPI run with more than one process?

What do you mean? Those “inactive with root” is there only to make sure that we gather some results across all MPI cores, like values of norms or eigenvalues. But there’s nothing to gather, for example, in case of Cholesky factorization.

**Comment by davydden** [comment]
@masterleinad 

> In reciprocal_condition_number the call to ppocon seems to require communication with an inactive MPI process.

does it work for you outside of this PR? I have a tester with ScaLAPACK and so far it did not pick up any deadlocks, also with Intel MKL: https://cdash.kyomu.43-1.org/buildSummary.php?buildid=14471 and https://cdash.kyomu.43-1.org/buildSummary.php?buildid=14478

To the best of my understanding, no communication should with processes outside the grid should be required. Since this PR does not change much functionality of ProcessGrid class, I don't think the problem is being introduce here. There are a few assumptions in `ProcessGrid` class, but they have Assert's to make sure it's the case, so I am a bit puzzled why would there be a problem. 

Another question: does it work for you with 4 mpi cores? In this case the grid is 2x2 and `mpi_communicator_inactive_with_root` should only have a single MPI process in it.

**Comment by masterleinad** [comment]
> does it work for you outside of this PR? I have a tester with ScaLAPACK and so far it did not pick up any deadlocks, also with Intel MKL: https://cdash.kyomu.43-1.org/buildSummary.php?buildid=14471 and https://cdash.kyomu.43-1.org/buildSummary.php?buildid=14478

Yes, the tests on the current master branch pass. I suspect the issue is really the different function called for `float`.

**Comment by masterleinad** [comment]
> Another question: does it work for you with 4 mpi cores? In this case the grid is 2x2 and mpi_communicator_inactive_with_root should only have a single MPI process in it.

No, I tried any number between 1 and 10. `scalapack_05` only passes when using one process.

**Comment by davydden** [comment]
> I suspect the issue is really the different function called for float.

Wow... what configuration do I have?

**Comment by masterleinad** [comment]
> Wow... what configuration do I have?

I am using `ScaLAPACK-2.0.2` and `clang-5.0.0` with `openmpi-2.0.1`.

**Comment by BenBrands** [comment]
I ran the ScaLAPACK tests on a different machine with ScaLAPACK-2.0.2, GCC-7.2.0 and OpenMPI-3.0.0 and everything passed. I also used numbers of MPI processes so that some processes did not participate in the ProcessGrid.

**Comment by davydden** [comment]
@masterleinad given the reply above, do u mind merging this PR?

**Comment by masterleinad** [comment]
`CDash` will test this with different configurations and tell us if not just my setup is weird.
Since this looks correct, we probably can't really do anything about this and works for you, I am fine with merging.

**Pull request merged**